### ISSUE TITLE
is captp pipelining result promises correctly?

### ISSUE THREAD

**Issue by warner** [Agoric/agoric-sdk#2846]
## Describe the bug

I'm still working on narrowing this down, but I think I'm seeing a problem with promise identity in the captp library. I'm working on a load generator that uses the fungible-faucet contract and simply taps the faucet on each cycle. I'm using an `agoric deploy` script like this:

```js
  async function faucetCycle3() {
    // make ourselves an invitation
    const invitation = E(creatorFacet).makeInvitation(); // result is rp-140
    // claim it
    const seatP = E(zoe).offer(invitation); // XX this isn't pipelining, arg is rp-142
    const paymentP = E(seatP).getPayout('Token');
    const payment = await paymentP;
    await E(bobPurse).deposit(payment);
    const amount = await E(bobPurse).getCurrentAmount();
    console.log(`new purse balance`, amount.value, new Date());
  }
```

When I look at the solo-to-chain traffic generated by this, the `makeInvitation()` message is sent with a result promise named `rp-140`. The following `offer()` message is supposed to use the same thing, but instead it emits `rp-142`, as if `invitation` is a brand new promise that hasn't been sent over the wire before. This results in a pipeline stall: the implementation of `offer` needs to wait for the might-be-a-Promise `invitation` argument to resolve before it can continue, but in this case the chain must wait for the solo machine to send the resolution notice. The full flow of messages is:

* solo->chain: `makeInvitation`, `offer`, `getPayout`
* chain->solo: resolve `invitation` (rp-140)
* solo->chain: resolve `invitation` (rp-142)
* chain->solo: resolve `seatP`, `paymentP`
* solo->chain: `deposit`
* chain->solo: resolve result
* solo->chain `getCurrentAmount`
* chain->solo: resolve `amount`

When I put this same code in a regular vat (using `SwingSet/test/test-liveslots.js` to watch the syscalls it produces), I see the `offer()` argument is exactly the same result promise that was attached to the `makeInvitation` message. Since the deploy script doesn't run in a vat, but rather in a JS environment with the captp library providing `E()`, I suspect that the capto library is behaving differently than liveslots.

@michaelfig does this sound like the way captp is supposed to behave w.r.t. result promises? I can put some time into producing a test case, but I'd like to start with just your educated guess as to what captp is intended to do.

**Comment by warner** [comment]
In captp, the result promise seems to be created with `makeQuestion`:

https://github.com/Agoric/agoric-sdk/blob/f686e589fbb477333790e9163c23671df231f54f/packages/captp/lib/captp.js#L166-L173

which doesn't look like it interacts with `valToSlot` at all. The corresponding code in SwingSet's `liveslots` is in `queueMessage`:

https://github.com/Agoric/agoric-sdk/blob/f686e589fbb477333790e9163c23671df231f54f/packages/SwingSet/src/kernel/liveSlots.js#L405-L410

which adds the return Promise to the tables just before giving it back to the caller:

https://github.com/Agoric/agoric-sdk/blob/f686e589fbb477333790e9163c23671df231f54f/packages/SwingSet/src/kernel/liveSlots.js#L429-L441

**Comment by michaelfig** [comment]
> @michaelfig does this sound like the way captp is supposed to behave w.r.t. result promises? I can put some time into producing a test case, but I'd like to start with just your educated guess as to what captp is intended to do.

Ahh, you've peeked behind the curtain.  `@agoric/captp` is a port/fork of an older liveslots informed by `rpc.capnp`.  All bugs are credited to me, and I have not kept it up-to-date wrt current liveslots.

I think your diagnosis is correct.  Would you be able to fix this if it is a blocker for you?  If it isn't a blocker, then it can wait.

**Comment by warner** [comment]
I don't have the bandwidth now, alas. It's not critical, but it's probably adding a 5-10 second delay to contract installs, more if we made a deliberate effort to improve how they're written (and take better advantage of pipelining.. the examples I copied the load generator from were using more `await` than would have been ideal). So it's probably the lowest-hanging fruit for a performance win on contract install and other frontend-driven chain interaction.
